src/transformer/lowercase.cpp

```diff
@@ -56,23 +56,23 @@ bool lowercase::needs_transform(std::string_view str)
         const __m128i cmp_lower = _mm_cmpgt_epi8(sse_mask_upper_bound, input_data); // Less than 'Z'
         const __m128i cmp_result =
             _mm_and_si128(cmp_upper, cmp_lower); // Combine the two comparison results
-        cmp_result_final = _mm_or_si128(cmp_result_final, cmp_result);
+        has_uppercase =
+            _mm_movemask_epi8(_mm_cmpeq_epi8(cmp_result, _mm_setzero_si128())) != 0xFFFF;
     }
 
-    bool is_lowercase =
-        _mm_movemask_epi8(_mm_cmpeq_epi8(cmp_result_final, _mm_setzero_si128())) == 0xFFFF;
-
-    for (std::size_t i = aligned_size; i < str.size(); i++) { is_lowercase &= !isupper(input[i]); }
+    for (std::size_t i = aligned_size; i < str.size() && !has_uppercase; i++) {
```

Comment:
it's not obvious that this will actually improve performance given that you're doing more work per loop and have more branches. Or, more exactly, that doing so will improve on average (it will likely improve for very long strings that have uppercase characters in the beginning). Did you run any benchmarks?